### ISSUE TITLE
AsParallel causes extremely slow 'WorkloadJitting' performance in BenchmarkDotNet benchmark

### ISSUE THREAD

**Issue by Treit** [dotnet/runtime#76689]
### Description

I managed to write a benchmark that makes the WorkloadJitting phase of BenchmarkDotNet perform extremely slowly.

For a count of 4:
```
WorkloadJitting  1: 1 op, 11779303800.00 ns, 11.7793 s/op
```

For a count of 8:
```
WorkloadJitting  1: 1 op, 29785615800.00 ns, 29.7856 s/op
```

Those are seconds!

The actual benchmark code runs fast after the WorkloadJitting phase finally completes.

I have included the minimal repro program in the Reproduction Steps section.

Note that removing the call to `AsParallel()` makes everything fast again.

### Reproduction Steps

```cs
using System;
using System.Collections.Generic;
using System.Linq;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

namespace Test
{
    public class PerfTest
    {
        [Params(4, 8)]
        public int PredicateCount { get; set; }

        [Benchmark]
        public void Test()
        {
            IEnumerable<int> rows = new int[] { 1234 };
            var predicates = new Func<int, bool>[PredicateCount];

            for (int i = 0; i < PredicateCount; i++)
            {
                predicates[i] = x => true;
            }

            foreach (var predicate in predicates)
            {
                rows = rows.AsParallel().Where(predicate);
            }

            var x = rows.ToList();
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            BenchmarkRunner.Run(typeof(Program).Assembly);
        }
    }
}
```

### Expected behavior

Should be fast.

### Actual behavior

Incredibly slow.

**Comment by EgorBo** [comment]
When I tried to profile this locally in VTune I got this:

![image](https://user-images.githubusercontent.com/523221/194173800-ca5008b3-51a6-4278-a8f8-16299a3fdb92.png)

**Comment by JulieLeeMSFT** [comment]
@Treit what is the .NET version you used?
Cc @dotnet/jit-contrib.

**Comment by Treit** [comment]
I tested on .NET 6. Also checked that it reproduces on .NET 5 as well. I did not test on .NET 7.

**Comment by Treit** [comment]
Here is an even simpler repro that is just a console program, not BenchmarkDotNet involved:

```cs
static void Main(string[] args)
{
    var sw = Stopwatch.StartNew();
    Console.WriteLine("About to run the test...");
    IEnumerable<int> rows = new int[] { 1234 };
    var predicates = new Func<int, bool>[4];

    for (int i = 0; i < 4; i++)
    {
        predicates[i] = x => true;
    }

    foreach (var predicate in predicates)
    {
        rows = rows.AsParallel().Where(predicate);
    }

    var x = rows.ToList();
    Console.WriteLine($"Test finished after {sw.ElapsedMilliseconds} ms.");
}
```

**Comment by Treit** [comment]
Here is the exact simplified code I used to reproduce this:
https://github.com/Treit/misc/tree/master/AsParallelIssue

![image](https://user-images.githubusercontent.com/8271812/194181183-688eefcb-4855-4540-82f9-f55747fa7422.png)

**Comment by stephentoub** [comment]
Your repro is calling AsParallel() N times as part of a single query.  Each one of those is going to end up demanding P threads, where by default P is the number of cores.  You're going to end up blocking all the thread pool's threads and then have to wait for the thread pool to slowly inject additional threads; once it's injected enough for all of those parallel queries to have their demands met, then everything instantly unblocks. If you change:
```C#
IEnumerable<int> rows = new int[] { 1234 };
```
to instead be:
```C#
ParallelQuery<int> rows = new int[] { 1234 }.AsParallel();
```
and then change:
```C#
rows = rows.AsParallel().Where(predicate);
```
to be:
```C#
rows = rows.Where(predicate);
```
you should see the problem go away.

**Comment by Treit** [comment]
So...instant thread pool exhaustion 😬

**Comment by Clockwork-Muse** [comment]
@stephentoub - granted you're not supposed to call it multiple times, but should we change `AsParallel()` to no-op on such calls?

**Comment by JulieLeeMSFT** [comment]
@Treit please try what stephentoub suggested and let us know if the issue goes away. Please close the issue if it does.

**Comment by stephentoub** [comment]
We could detect and nop the AsParallel in _some_ cases, e.g. this exact repro could be made to "just work"... we'd just add a check to `AsParallel()` that was like `if (source is ParallelQuery<T> pq) return pq;`.  That would, however, only work if it's actually being called on such a thing.  But if instead of your repro, which is effectively:
```C#
rows
    .AsParallel().Where(predicate)
    .AsParallel().Where(predicate)
    .AsParallel().Where(predicate)
    .AsParallel().Where(predicate);
```
the developer had the equivalent of:
```C#
rows
    .AsParallel().Where(predicate).AsSequential().Where(predicate)
    .AsParallel().Where(predicate).AsSequential().Where(predicate)
    .AsParallel().Where(predicate).AsSequential().Where(predicate)
    .AsParallel().Where(predicate).AsSequential().Where(predicate);
```
it would not help with that case, nor do I see a good way to make that viable.  And while that construction looks strange and like "who would write that", both of these would likely arise not because the developer explicitly wrote such code but because it happened due to composition, e.g.
```C#
static IEnumerable<int> GetProcessor() => _source.AsParallel().Where(...);
static IEnumerable<int> GetLayeredProcessor() => GetProcessor().Where(...);
...
IEnumerable<int> e = GetLayeredProcessor().AsParallel().Where(...);
... // etc.
```

I'd be ok if we wanted to add a condition like I cited, but I don't know if it would really move the needle.

**Comment by rgroenewoudt** [comment]
What about a code analyzer that detects multiple AsParallel() calls on the same variable, including loops?

**Comment by stephentoub** [comment]
> What about a code analyzer that detects multiple AsParallel() calls on the same variable, including loops?

Is this so common a mistake?  An analyzer might be possible, but I'm not sure it's worth the effort or cycles.

**Comment by stephentoub** [comment]
> We could detect and nop the AsParallel in some cases, e.g. this exact repro could be made to "just work"... we'd just add a check to AsParallel() that was like

I took a look, and this would actually be a breaking change.  Some query operators, like WithExecutionMode, are only usable once in a query; AsParallel can thus be used to essentially start a new query, but if we made it a nop in these cases, then such uses could start causing those single-use operators to fail.

I'm going to close this as by-design.  Thanks for the discussion.
